src/kubernetes.rs

```diff
@@ -0,0 +1,182 @@
+    let mut url = Url::parse("https://something/api/v1/").unwrap();
+    url.path_segments_mut()
+        .unwrap()
+        .push("namespaces")
+        .push(namespace)
+        .push("pods")
+        .push(pod_name);
+    let url = url.path();
+    let result = try_with!(
+        Command::new("kubectl")
+            .arg("get")
+            .arg("--raw")
+            .arg(url)
+            .output(),
+        "kubctl command cannot be spawned"
+    );
+    if !result.status.success() {
+        let stderr = String::from_utf8_lossy(&result.stderr);
+        bail!(
+            "kubectl get pod request failed (ret code {:?}): {}",
+            result.status.code(),
+            stderr
+        );
+    }
+
+    // parse response. Spec:
+    // https://kubernetes.io/docs/reference/kubernetes-api/workload-resources/pod-v1/#Pod
+    let json: json::Value = try_with!(
+        json::from_str(try_with!(
+            from_utf8(&result.stdout),
+            "response contains non-utf8"
+        )),
+        "failed to parse kubectl get pod response"
+    );
+    let containers = require_with!(
+        json["status"]["containerStatuses"].as_array(),
+        "failed to parse kubectl get pod response json 1"
+    );
+    if containers.is_empty() {
+        bail!("no containers present to attach to");
+    }
+    // choose container according to parse_userinput
+    let container = match container_name {
+        None => &containers[0],
+        Some(name) => require_with!(
+            containers.iter().find(|status| status["name"] == name),
+            "no container {} found in {}/{}",
+            name,
+            namespace,
+            pod_name
+        ),
+    };
+    let containerid = require_with!(
+        container["containerID"].as_str(),
+        "failed to parse kubectl get pod response json 2"
+    );
+    let containerid = require_with!(
+        containerid.strip_prefix("containerd://"),
+        "unexpected/unparsable containerd id"
+    );
+    Ok(String::from(containerid))
+}
+
+pub fn find_cgroup(containerdid: String) -> Result<PathBuf> {
+    let path = visit_dirs(
+        &PathBuf::from("/sys/fs/cgroup"),
+        &OsString::from(containerdid),
+    )?;
+    Ok(path)
+}
+
+// one possible implementation of walking a directory from
+// https://doc.rust-lang.org/std/fs/fn.read_dir.html
+fn visit_dirs(dir: &Path, containerdid: &OsString) -> Result<PathBuf> {
+    for entry in try_with!(std::fs::read_dir(dir), "cannot list {:?}", dir) {
+        let entry = try_with!(entry, "cannot read entry in dir {:?}", dir);
+        if &entry.file_name() == containerdid {
+            return Ok(entry.path());
+        }
+        let path = entry.path();
+        if path.is_dir() {
+            if let Ok(path) = visit_dirs(&path, containerdid) {
+                return Ok(path);
+            }
+        }
+    }
+    bail!("nothing found");
+}
+
+/// return any pid part of this cgroup
```

Comment:
I suppose pid 1 will be the first pid in this cgroup, which is also what we usually want.

Reply:
> cgroup.procs: list of thread group IDs in the cgroup. This list is **not guaranteed to be sorted** or free of duplicate TGIDs, and userspace should sort/uniquify the list if this property is required. Writing a thread group ID into this file moves all threads in that group into this cgroup.

from https://www.kernel.org/doc/html/latest/admin-guide/cgroup-v1/cgroups.html

Do we strictly require the cgroup's "root" pid?

Reply:
The root pid has likely the most privileges. So eventually it would be better to have it.